vaadin-flow-components-shared-parent/vaadin-flow-components-base/src/main/java/com/vaadin/flow/component/HasSelection.java

```diff
@@ -0,0 +1,126 @@
+     */
+    default void setSelectionRange(int selectionStart, int selectionEnd) {
+        getElement().executeJs("var origType = this.inputElement.type;"
+                + "var needsTypeHack = (origType == 'number' || origType == 'email');"
+                + "if(needsTypeHack) {this.inputElement.type = 'text';}"
+                + "this.inputElement.setSelectionRange($0,$1);"
+                + "if(needsTypeHack) {this.inputElement.type = origType;};"
+                + "this.inputElement.focus();", selectionStart, selectionEnd);
+    }
+
+    /**
+     * Sets the cursor position to given index.
+     *
+     * @param cursorPosition
+     *            the cursor position
+     */
+    default void setCursorPosition(int cursorPosition) {
+        setSelectionRange(cursorPosition, cursorPosition);
+    }
+
+    interface SelectionRangeCallback extends Serializable {
+        /**
+         * This method is called with the current selection of the field.
+         *
+         * @param start
+         *            the start of the selection (inclusive)
+         * @param end
+         *            the end of the selection (inclusive)
+         * @param content
+         *            the string content currently selected
+         */
+        void selectionRange(int start, int end, String content);
+    }
+
+    /**
+     * Asynchronously gets the current selection for this field.
+     *
+     * @param callback
+     *            the callback to notify the selection
+     */
+    default void getSelectionRange(SelectionRangeCallback callback) {
+        getElement().executeJs("" + "var res = {};"
+                + "var origType = this.inputElement.type;"
+                + "var needsTypeHack = (origType == 'number' || origType == 'email');"
+                + "if(needsTypeHack) {this.inputElement.type = 'text';}"
+                + "res.start = this.inputElement.selectionStart;"
+                + "res.end = this.inputElement.selectionEnd;"
+                + "res.content = this.inputElement.value.substring(res.start, res.end);"
+                + "if(needsTypeHack) {this.inputElement.type = origType;}"
+                + "return res;").then(jsonValue -> {
+                    if (jsonValue instanceof JsonObject) {
+                        JsonObject jso = (JsonObject) jsonValue;
+                        callback.selectionRange((int) jso.getNumber("start"),
+                                (int) jso.getNumber("end"),
+                                jso.getString("content"));
+                    }
+                });
+    }
+
+    /**
+     * Asynchronously gets the current cursor position for this field.
+     *
+     * @param callback
+     *            the callback to notify the position
+     */
+    default void getCursorPosition(SerializableConsumer<Integer> callback) {
+        this.getSelectionRange((start, e, c) -> {
+            callback.accept(start);
+        });
+    }
```

Comment:
Insert/replace text on the selection. Like variable values when editing some template or think of creating a link around selected text in an html or markdown editor.

Everything asynchronous can always have timing issues. Those you presented are though rather hypothetical as those are most often initiated by a user who expect something to happen (e.g. insert link button).

Reply:
Let's say an application has a keyboard shortcut for inserting text on the selection. To have it work as expected, the user would need to know to wait for a second after typing the shortcut so that the cursor position wouldn't change during the roundtrips.

This may be even a bigger issue when you use the selection for some operation on the server-side, something that may not provide direct feedback to the user. In that case, the app logic could work with invalid values in case the user doesn't keep long enough pauses.

The reliability of this API depends on the network latency and we can't predict all use cases people may come up with.

Reply:
If a user uses a keyboard shortcut (which is already a special case), they most likely know what he is doing and expect that certain thing will happen, and waits for it to happen. If they accidentally use the click shortcut, then random things will happen, regardless of how much they happen to modify the content during the server round trip. But yeah, we can't predict with asynchronous API that what random things will happen.

BTW. I think we shouldn't talk about _a second_, even though you didn't mean that. A typical round trip is less than 100ms.

Reply:
Oh, yeah, there it might have, as there logic would be initiated with a button or similar.

Reply:
But still, rather hypothetical situation that could happen already (e.g. another thread/user changing value in UI). And the Javadocs and method signature rather well states that this happens asynchronously.

Reply:
Should `getCursorPosition()` then also be marked as `@Deprecated` since it requires the user to acknowledge the reliability issues with the API?

BTW The insert/replace text on the selection case would probably be more reliable if there was a selection _listener_ instead of a getter.

Reply:
A listener would indeed have less timing issues, but then we'd be bombarding the server on each selection changes, which would not be very optimal regarding client-server communication. If that is level of reliability is required for the use case, a component with client-side logic is the only sane way.

Why is getCursorPosition different than getSelectionRange?

Reply:
In number fields that has issues with Chrome in the same way as getSelectionRange? Yes.

Reply:
> Why is getCursorPosition different than getSelectionRange?

No difference from the reliability point of view. Should apply to both then.

Reply:
Assuming you started to discuss about the Chrome issues: deprecated also getCursorPosition now in the three fields where Chrome bugs. Also added a reason for deprecation.

Reply:
since attribute should actually contain version, so abusing it here. I'll remove those...

Reply:
✅

Reply:
> A listener would indeed have less timing issues, but then we'd be bombarding the server on each selection changes, which would not be very optimal regarding client-server communication.

This issue could be mitigated by debouncing the selection events, similarly to `textField.setValueChangeMode(...)`

Reply:
If we'd add `setTrackSelection` method it could always work without asynchronous API 🤔

Reply:
Here's a recording of the issue discussed in this thread

https://user-images.githubusercontent.com/1222264/175267161-31d50196-b298-4eb8-8d0b-556657216655.mp4

Reply:
Yes, as discussed, if you have latency calculated in seconds and not waiting for the action you requested to happen, there will be timing issues with asynchronous API. Don't know what is your point.

Reply:
If we'd add the selection detail on each value change and fix event ordering bugs and make it configurable  with a method like `setTrackSelection` it might be better. But with large latency one might still for example change the text length before the "target location", before the new text arrives.

There are a lot of other bugs in existing features and improvements we should probably focus more than this. The world will never be perfect.

Reply:
> Don't know what is your point.

As suggested above, we could probably design the API to work more reliably (perhaps a lazy selection listener) so that it would cause less frustration for our users.

Reply:
Who wants to listen for selection events, lazily or not 🤔

Reply:
That was also my initial question about the suggested getter APIs. And I guess the answer is also the same. 

But I mentioned the listener as just one suggestion, it may not be optimal either from DX perspective, but neither is an asynchrnous getter, clearly. One solution could be to make the current selection a (lazily synchronized) server-side property of the component. I wonder how it worked in V8?

Reply:
I don't think the answer is same. I can't come up with a requirement when one would like to act when selection of text changes, but one wants to read that for sure in other listeners.

The alternative solution would be to have synchronous API and transfer the selection details all the time (at least on each server round trip). But that would have other issues (more complicated implementation, current framework bugs, increased data transfer and/or trowing runtime exceptions if developer hasn't explicitly requested to synchronise them), and the solution would not solve all timing issues you were concerned (but would make the 50% less likely, 1 server round trip instead of 2).

As I see this is that we can either amuse ourselves by making this slightly better with a lot of work or to ship it now and something more important instead. As it (getSelectionRange()) is rather rarely needed feature, getting a decent solution with a smaller price would make sense here IMO. Especially as the alternative implementation can be shipped in a rather backwards compatible manner (synchronous method would have different signature).

Reply:
> I don't think the answer is same. I can't come up with a requirement when one would like to act when selection of text changes, but one wants to read that for sure in other listeners.

The point was that using the listener approach, replacing selection (which was the use case) works more reliably than with the async getter. Didn't mean that I have other specific use cases in mind for a text selection listener specifically.

> I wonder how it worked in V8?

To answer myself, the `getCursorPosition()` in V8 is synchronous (returns an `int`), so the state of the cursor position _is_ synchronized server-side. To match the API of V8 TextField, we could take the approach of synchronizing the selection state and then providing getters for it.

![Screenshot 2022-06-23 at 15 07 16](https://user-images.githubusercontent.com/1222264/175295646-50d6987f-cf65-4668-b550-a5653a987638.png)

Reply:
It might have been me implementing that ages ago. For Vaadin  < 10, that was rather easy to implement as a side effect when synchronising the textfield value, for this new beast, I have no idea how to do it efficiently.

Reply:
BTW. If this rather fundamental bug in Flow isn't fixed, the "V8 style" solution isn't going to work too well. https://github.com/vaadin/flow/pull/14035